src/ui/public/visualize/components/visualization_chart.js

```diff
@@ -85,23 +85,21 @@ export class VisualizationChart extends Component {
   };
 
   componentDidMount() {
-    const { vis, listenOnChange } = this.props;
+    const { vis, onInit } = this.props;
     const Visualization = vis.type.visualization;
 
     this.visualization = new Visualization(this.chartDiv.current, vis);
 
-    if (this.visualization.isLoaded) {
-      this.visualization.isLoaded().then(() => {
-        vis.initialized = true;
-      });
-    } else {
-      vis.initialized = true;
+    if (onInit) {
+      if (this.visualization.isLoaded) {
```

Comment:
Since we currently don't have static typing in this part, I see how visualization implementations could easily try to use `onInit` to just initialize and forgetting to return us a promise here. Could we perhaps change that code to:

```js
if (onInit) {
  const visLoaded = this.visualization.isLoaded ? this.visualization.isLoaded() : true;
  Promise.resolve(visLoaded).then(onInit);
}
```